### PR TITLE
8324665: Loose matching of space separators in the lenient date/time parsing mode

### PR DESCRIPTION
**Pull request by naotoj**
Implementing "loose matching" of space separators in both `java.time.format.DateTimeFormatter` and `java.text.DateFormat` on lenient parsing. This will effectively fix the NNBSP issues on parsing time with am/pm markers introduced with CLDR version 42 (https://inside.java/2023/03/28/quality-heads-up/). A draft CSR has also been drafted.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires CSR request [JDK-8324990](https://bugs.openjdk.org/browse/JDK-8324990) to be approved

### Issues
 * [JDK-8324665](https://bugs.openjdk.org/browse/JDK-8324665): Loose matching of space separators in the lenient date/time parsing mode (**Enhancement** - P4)
 * [JDK-8324990](https://bugs.openjdk.org/browse/JDK-8324990): Loose matching of space separators in the lenient date/time parsing mode (**CSR**)

### Reviewers
 * [Joe Wang](https://openjdk.org/census#joehw) (@JoeWang-Java - **Reviewer**) ⚠️ Review applies to [0c950941](https://git.openjdk.org/jdk/pull/17678/files/0c950941bf6beca18cc1e42a2ca0c7f7abc0a05e)
 * [Justin Lu](https://openjdk.org/census#jlu) (@justin-curtis-lu - Committer) ⚠️ Review applies to [0c950941](https://git.openjdk.org/jdk/pull/17678/files/0c950941bf6beca18cc1e42a2ca0c7f7abc0a05e)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17678/head:pull/17678` \
`$ git checkout pull/17678`

Update a local copy of the PR: \
`$ git checkout pull/17678` \
`$ git pull https://git.openjdk.org/jdk.git pull/17678/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17678`

View PR using the GUI difftool: \
`$ git pr show -t 17678`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17678.diff">https://git.openjdk.org/jdk/pull/17678.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17678#issuecomment-1922460612)

### CODE REVIEW

**Comment by JoeWang-Java** [comment]
The compatibility risk is low for this change itself as assessed in the CSR, and this change looks good to me. However, just would like to note that it conforms to CLDR, but doesn't address completely the issue reported in JDK-8304925. Since the default parsing mode is strict in java.time.format, applications would still have to make code changes when moving existing code to the new JDK releases. In java.text then, all space separators will be parsed by default, that would be a welcoming behavior change to me.

**Comment by naotoj** [comment]
> Since the default parsing mode is strict in java.time.format, applications would still have to make code changes when moving existing code to the new JDK releases.

You are right, Joe. Apps using `java.time` formatters would still need to make changes to their code base. However, the bar is significantly lower. They would have to write some ugly workaround without this fix.

**Review by JoeWang-Java — APPROVED** [review]

**Comment by JoeWang-Java** [comment]
> > Since the default parsing mode is strict in java.time.format, applications would still have to make code changes when moving existing code to the new JDK releases.
> 
> You are right, Joe. Apps using `java.time` formatters would still need to make changes to their code base. However, the bar is significantly lower. They would have to write some ugly workaround without this fix.

True, indeed what an improvement / enhancement is for.

File: src/java.base/share/classes/java/text/DateFormat.java [hunk 1]
```diff
@@ -746,6 +746,10 @@ public TimeZone getTimeZone()
      * <p>This leniency value is overwritten by a call to {@link
      * #setCalendar(java.util.Calendar) setCalendar()}.
      *
+     * @implSpec A {@link Character#SPACE_SEPARATOR SPACE_SEPARATOR} in the input
+     * text will match any other {@link Character#SPACE_SEPARATOR SPACE_SEPARATOR}s
+     * in the pattern with the lenient parsing; otherwise, it will not match.
```
Comment (justin-curtis-lu) [inline src/java.base/share/classes/java/text/DateFormat.java:751]: LGTM.

Nit: Might read better as either "in the pattern with lenient parsing" or "in the pattern when parsing is lenient"

**Review by justin-curtis-lu — APPROVED** [review]

**Comment by RogerRiggs** [comment]
> You are right, Joe. Apps using `java.time` formatters would still need to make changes to their code base. However, the bar is significantly lower. They would have to write some ugly workaround without this fix.

I'd also expect that applications accepting user input may already be using lenient mode to make the application easier to use; so in come cases, it will come without new code changes.

File: src/java.base/share/classes/java/text/DateFormat.java (hunk 1 above)
Comment (naotoj) [reply src/java.base/share/classes/java/text/DateFormat.java]: Thanks. Fixed.

**Pull request closed without merging**